eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelayer.scala

```diff
@@ -48,34 +51,47 @@ object NodeRelayer {
     Behaviors.setup { context =>
       Behaviors.withMdc(Logs.mdc(category_opt = Some(Logs.LogCategory.PAYMENT)), mdc) {
         Behaviors.receiveMessage {
           case Relay(nodeRelayPacket) =>
-            import nodeRelayPacket.add.paymentHash
-            children.get(paymentHash) match {
-              case Some(handler) =>
-                context.log.debug("forwarding incoming htlc to existing handler")
-                handler ! NodeRelay.Relay(nodeRelayPacket)
-                Behaviors.same
+            val htlcIn = nodeRelayPacket.add
+            nodeRelayPacket.outerPayload.paymentSecret match {
+              case Some(paymentSecret) =>
+                val childKey = PaymentKey(htlcIn.paymentHash, paymentSecret)
+                children.get(childKey) match {
+                  case Some(handler) =>
+                    context.log.debug("forwarding incoming htlc #{} from channel {} to existing handler", htlcIn.id, htlcIn.channelId)
+                    handler ! NodeRelay.Relay(nodeRelayPacket)
+                    Behaviors.same
+                  case None =>
+                    val relayId = UUID.randomUUID()
+                    context.log.debug(s"spawning a new handler with relayId=$relayId")
+                    val handler = context.spawn(NodeRelay.apply(nodeParams, context.self, register, relayId, nodeRelayPacket, childKey.paymentSecret, outgoingPaymentFactory), relayId.toString)
+                    context.log.debug("forwarding incoming htlc #{} from channel {} to new handler", htlcIn.id, htlcIn.channelId)
+                    handler ! NodeRelay.Relay(nodeRelayPacket)
+                    apply(nodeParams, register, outgoingPaymentFactory, children + (childKey -> handler))
+                }
               case None =>
-                val relayId = UUID.randomUUID()
-                context.log.debug(s"spawning a new handler with relayId=$relayId")
-                val outgoingPaymentFactory = NodeRelay.SimpleOutgoingPaymentFactory(nodeParams, router, register)
-                val handler = context.spawn(NodeRelay.apply(nodeParams, context.self, register, relayId, paymentHash, outgoingPaymentFactory), relayId.toString)
-                context.log.debug("forwarding incoming htlc to new handler")
-                handler ! NodeRelay.Relay(nodeRelayPacket)
-                apply(nodeParams, router, register, children + (paymentHash -> handler))
+                context.log.warn("rejecting htlc #{} from channel {}: missing payment secret", htlcIn.id, htlcIn.channelId)
+                val failureMessage = IncorrectOrUnknownPaymentDetails(htlcIn.amountMsat, nodeParams.currentBlockHeight)
+                val cmd = CMD_FAIL_HTLC(htlcIn.id, Right(failureMessage), commit = true)
+                PendingRelayDb.safeSend(register, nodeParams.db.pendingRelay, htlcIn.channelId, cmd)
```

Comment:
I think it would make sense, yes. It's been introduced in December 2019, so we can probably consider them mandatory now.
When we make that change, we should set the feature bit as mandatory. I added that to my TODO-list.